go/vt/vtctl/internal/grpcshim/bidi_stream.go

```diff
@@ -0,0 +1,186 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpcshim
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// ErrStreamClosed is the error types embedding BidiStream should return when
+// their Send method is called and IsClosed returns true.
+var ErrStreamClosed = errors.New("stream closed for sending")
+
+// BidiStream is a shim struct implementing both the grpc.ClientStream and
+// grpc.ServerStream interfaces. It can be embedded into other types that need
+// all of those methods to satisfy the compiler, but are only interested in the
+// parameterized Send/Recv methods typically called by gRPC streaming servers
+// and clients. For example, in the localvtctldclient:
+//
+// 	type backupStreamAdapter struct {
+// 		*grpcshim.BidiStream
+// 		ch chan *vtctldatapb.BackupResponse
+// 	}
+//
+// 	func (stream *backupStreamAdapter) Recv() (*vtctldatapb.BackupResponse, error) {
+// 		select {
+// 		case <-stream.Context().Done():
+// 			return nil, stream.Context().Err()
+// 		case <-stream.Closed():
+// 			// Stream has been closed for future sends. If there are messages that
+// 			// have already been sent, receive them until there are no more. After
+// 			// all sent messages have been received, Recv will return the CloseErr.
+// 			select {
+// 			case msg := <-stream.ch:
+// 				return msg, nil
+// 			default:
+// 				return nil, stream.CloseErr()
+// 			}
+// 		case err := <-stream.ErrCh:
+// 			return nil, err
+// 		case msg := <-stream.ch:
+// 			return msg, nil
+// 		}
+// 	}
+//
+// 	func (stream *backupStreamAdapter) Send(msg *vtctldatapb.BackupResponse) error {
+// 		select {
+// 		case <-stream.Context().Done():
+// 			return stream.Context().Err()
+//		case <-stream.Closed():
+//			return grpcshim.ErrStreamClosed
+// 		case stream.ch <- msg:
+// 			return nil
+// 		}
+// 	}
+//
+// 	// Backup is part of the vtctlservicepb.VtctldClient interface.
+// 	func (client *localVtctldClient) Backup(ctx context.Context, in *vtctldatapb.BackupRequest, opts ...grpc.CallOption) (vtctlservicepb.Vtctld_BackupClient, error) {
+// 		stream := &backupStreamAdapter{
+// 			BidiStream: grpcshim.NewBidiStream(ctx),
+// 			ch:         make(chan *vtctldatapb.BackupResponse, 1),
+// 		}
+// 		go func() {
+// 			err := client.s.Backup(in, stream)
+// 			stream.CloseWithError(err)
+// 		}()
+// 		return stream, nil
+// 	}
+//
+type BidiStream struct {
+	// ErrCh receives errors mid-stream, and should be selected on with the
+	// same priority as stream.ch and stream/send contexts' cancellations in
+	// Recv methods.
+	ErrCh chan error
+
+	ctx              context.Context
+	sendClosedCtx    context.Context
+	sendClosedCancel context.CancelFunc
+
+	m        sync.RWMutex
```

Comment:
Quoting @sougou
" Programmatic RWLocks are different from the equivalent db functionality. They're actually more expensive than a normal mutex. They're only useful if you want multiple readers to hold them concurrently and for long periods, and then occasionally want to use a write lock to change the values. Also, mixing RW locking with write locks can easily lead to deadlocks if one doesn't think through the scenarios carefully.
So, in most cases, it's simpler and more efficient to use shortly held write locks even if RW lock seems functionally more appropriate,"

Reply:
Following up from our other conversation:

> looking again at the code, we don't really have many concurrent readers. Only if you passed the stream shim to multiple goroutines, which we currently don't do and probably never would do in practice. I'll switch to a plain-old Mutex